File: 0006-zigzag-conversion/0006-zigzag-conversion.cpp
class Solution {
public:
    string convert(string input, int numRows) {

        if (numRows == 1 || numRows >= (int)input.length()) return input;

    string result;
    result.reserve(input.length()); // reserve capacity for efficiency

    for (int row = 0; row < numRows; row++) {
        int currentIndex = row;
        int stepDown = 2 * (numRows - 1 - row);
        int stepUp = 2 * row;
        bool movingDown = true;

        while (currentIndex < (int)input.length()) {
            result += input[currentIndex];

            if (row == 0) {
                currentIndex += stepDown;
            } else if (row == numRows - 1) {
                currentIndex += stepUp;
            } else {
                if (movingDown) {
                    currentIndex += stepDown;
                } else {
                    currentIndex += stepUp;
                }
                movingDown = !movingDown;
            }
        }
    }

    return result;
        
    }
};

File: 0006-zigzag-conversion/0006-zigzag-conversion.java
class Solution {
    public String convert(String s, int nr) {

        if(nr == 1 || nr >= s.length()){
            return s;
        }

        StringBuilder[] rb = new StringBuilder[nr];

        for(int i=0;i <nr; i++){
            rb[i] = new StringBuilder();
        }

        int cr = 0;
        boolean md = true;

        for(char cc : s.toCharArray()){
            rb[cr].append(cc);

            if(cr == 0){
                md = true;
            }else if(cr == nr -1){
                md = false;
            }

            cr += md ? 1 : -1;
        }

        StringBuilder fr = new StringBuilder();
        for(StringBuilder ra : rb){
            fr.append(ra);
        }

        return fr.toString();
        
    }
}